analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs

```diff
@@ -179,8 +184,71 @@ private ProgramState RemoveLock(SymbolicContext context, ISymbol symbol)
+                                                                               KnownType type,
+                                                                               int parameterCount,
+                                                                               int refParameterIndex,
+                                                                               string lockObjectParameterName,
+                                                                               string refParameterName,
+                                                                               params string[] methodNames) =>
+            context.Operation.Instance.AsInvocation() is { } invocation
+            && IsInvocationWithBoolRefParam(invocation, type, parameterCount, refParameterIndex, methodNames)
+            && ArgumentSymbol(invocation, 0, lockObjectParameterName) is { } lockObject
+            && ArgumentSymbol(invocation, refParameterIndex, refParameterName) is { } refParameter
+                ? new RefParamContext(context, lockObject, refParameter)
+                : null;
+
+        private static RefParamContext InvocationWithBoolRefParam(SymbolicContext context,
+                                                                  KnownType type,
+                                                                  int parameterCount,
+                                                                  int refParameterIndex,
+                                                                  string refParameterName,
+                                                                  params string[] methodNames) =>
+            context.Operation.Instance.AsInvocation() is { } invocation
+            && IsInvocationWithBoolRefParam(invocation, type, parameterCount, refParameterIndex, methodNames)
+            && invocation.Instance.TrackedSymbol() is { } lockObject
+            && ArgumentSymbol(invocation, refParameterIndex, refParameterName) is { } refParameter
+                ? new RefParamContext(context, lockObject, refParameter)
+                : null;
+
+        private static bool IsInvocationWithBoolRefParam(IInvocationOperationWrapper invocation, KnownType type, int parameterCount, int refParameterIndex, params string[] methodNames) =>
+            invocation.TargetMethod.IsAny(type, methodNames)
+            && invocation.TargetMethod.Parameters.Length == parameterCount
+            && invocation.TargetMethod.Parameters[refParameterIndex].IsType(KnownType.System_Boolean);
+
+        private ProgramState[] HeldAndNotHeldStates(RefParamContext context) =>
+            new[]
+            {
+                AddLock(context.SymbolicContext, context.LockSymbol).SetSymbolConstraint(context.RefParamSymbol, context.SymbolicContext.SymbolicValueCounter, BoolConstraint.True),
+                context.SymbolicContext.State.SetSymbolConstraint(context.RefParamSymbol, context.SymbolicContext.SymbolicValueCounter, BoolConstraint.False)
```

Comment:
Let's extract the duplicated stuff into a method that will get the bool constraint and state that it should apply it to.
```suggestion
                context.SetRefConstraint(BoolConstraint.True, AddLock(context.SymbolicContext, context.LockSymbol)),
                context.SetRefConstraint(BoolConstraint.False, SymbolicContext.State)
```

Reply:
My suggestion started with the `BoolConstraint` argument, because it makes the two invocations lines more table-like and easier to read and see the difference.